abnormal behavior when resuming a stopped changefeed

## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.

- start two captures
- start a changefeed
- stop the changefeed via `curl -X POST -d "admin-job=3&cf-id=xxx" http://127.0.0.1:8301/capture/owner/admin`
- resume the changefeed (in fact resume should make no sense) via `curl -X POST -d "admin-job=2&cf-id=xxx" http://127.0.0.1:8301/capture/owner/admin`, send HTTP request to both owner capture and non-owner capture

2. What did you expect to see?

returns some expected error messages

3. What did you see instead?

non owner node
```
2020-05-08 11:53:48.360571 I | http: panic serving 127.0.0.1:33912: runtime error: invalid memory address or nil pointer dereference
goroutine 1489 [running]:
net/http.(*conn).serve.func1(0xc000354780)
        net/http/server.go:1767 +0x139
panic(0x1d9d8c0, 0x3af1740)
        runtime/panic.go:679 +0x1b2
sync.(*Mutex).Lock(...)
        sync/mutex.go:74
github.com/pingcap/ticdc/cdc.(*Owner).EnqueueJob(0x0, 0xc00056a012, 0x24, 0x2, 0xc002dfc4c0, 0x0)
        github.com/pingcap/ticdc@/cdc/owner.go:589 +0x7c
github.com/pingcap/ticdc/cdc.(*Server).handleChangefeedAdmin(0xc0000ac5a0, 0x247c660, 0xc0000b0a80, 0xc002f08d00)
        github.com/pingcap/ticdc@/cdc/http_handler.go:100 +0x1ba
net/http.HandlerFunc.ServeHTTP(0xc00034e8a0, 0x247c660, 0xc0000b0a80, 0xc002f08d00)
        net/http/server.go:2007 +0x44
net/http.(*ServeMux).ServeHTTP(0xc0000a2980, 0x247c660, 0xc0000b0a80, 0xc002f08d00)
        net/http/server.go:2387 +0x1bd
net/http.serverHandler.ServeHTTP(0xc0000ec000, 0x247c660, 0xc0000b0a80, 0xc002f08d00)
        net/http/server.go:2802 +0xa4
net/http.(*conn).serve(0xc000354780, 0x2484aa0, 0xc0007b7980)
        net/http/server.go:1890 +0x875
created by net/http.(*Server).Serve
        net/http/server.go:2927 +0x38e
```

owner node
```
[2020/05/08 11:55:37.657 +08:00] [INFO] [owner.go:418] ["handle admin job"] [changefeed=3f4c9771-46fd-4ed7-a6e5-31247d74f7bf] [type="resume changefeed"]
[2020/05/08 11:55:37.658 +08:00] [ERROR] [owner.go:526] ["owner exited with error"] [error="query detail id 3f4c9771-46fd-4ed7-a6e5-31247d74f7bf: changefeed not exists"] [errorVerbose="changefeed not exists\nquery detail id 3f4c9771-46fd-4ed7-a6e5-31247d74f7bf\ngithub.com/pingcap/ticdc/cdc/kv.CDCEtcdClient.GetChangeFeedInfo\n\tgithub.com/pingcap/ticdc@/cdc/kv/etcd.go:140\ngithub.com/pingcap/ticdc/cdc.(*Owner).handleAdminJob\n\tgithub.com/pingcap/ticdc@/cdc/owner.go:452\ngithub.com/pingcap/ticdc/cdc.(*Owner).run\n\tgithub.com/pingcap/ticdc@/cdc/owner.go:565\ngithub.com/pingcap/ticdc/cdc.(*Owner).Run\n\tgithub.com/pingcap/ticdc@/cdc/owner.go:523\ngithub.com/pingcap/ticdc/cdc.(*Server).run.func2\n\tgithub.com/pingcap/ticdc@/cdc/server.go:169\nruntime.goexit\n\truntime/asm_amd64.s:1357"]
```

4. Versions of the cluster

    - TiCDC version (execute `cdc version`):

```
cdc version
Release Version: v4.0.0-rc.1-12-g39e75c3
Git Commit Hash: 39e75c3524e2224715fb162aa84a380dc40cf2e6
Git Branch: master
UTC Build Time: 2020-05-08 03:58:42
Go Version: go version go1.13.3 linux/amd64
```

--- a/cdc/kv/etcd.go
+++ b/cdc/kv/etcd.go
@@ -509,6 +509,32 @@ func (c CDCEtcdClient) PutChangeFeedStatus(
 	return errors.Trace(err)
 }
 
+// SetChangeFeedStatusTTL sets the TTL of changefeed synchronization status
+func (c CDCEtcdClient) SetChangeFeedStatusTTL(
+	ctx context.Context,
+	changefeedID string,
+	ttl int64,
+) error {
+	key := GetEtcdKeyJob(changefeedID)
+	leaseResp, err := c.Client.Grant(ctx, ttl)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	status, _, err := c.GetChangeFeedStatus(ctx, changefeedID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	statusStr, err := status.Marshal()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	_, err = c.Client.Put(ctx, key, statusStr, clientv3.WithLease(leaseResp.ID))
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return errors.Trace(err)
+}
+
 // PutAllChangeFeedStatus puts ChangeFeedStatus of each changefeed into etcd
 func (c CDCEtcdClient) PutAllChangeFeedStatus(ctx context.Context, infos map[model.ChangeFeedID]*model.ChangeFeedStatus) error {
 	var (

--- a/cdc/kv/etcd_test.go
+++ b/cdc/kv/etcd_test.go
@@ -261,14 +261,14 @@ func (s *etcdSuite) TestSetChangeFeedStatusTTL(c *check.C) {
 	c.Assert(status, check.DeepEquals, &model.ChangeFeedStatus{
 		ResolvedTs: 1,
 	})
-	err = s.client.SetChangeFeedStatusTTL(ctx, "test1", 2 /* seconds */)
+	err = s.client.SetChangeFeedStatusTTL(ctx, "test1", 1 /* seconds */)
 	c.Assert(err, check.IsNil)
 	status, _, err = s.client.GetChangeFeedStatus(ctx, "test1")
 	c.Assert(err, check.IsNil)
 	c.Assert(status, check.DeepEquals, &model.ChangeFeedStatus{
 		ResolvedTs: 1,
 	})
-	time.Sleep(3 * time.Second)
+	time.Sleep(1.5 * time.Second)
 	_, _, err = s.client.GetChangeFeedStatus(ctx, "test1")
 	c.Assert(errors.Cause(err), check.Equals, model.ErrChangeFeedNotExists)
 }

--- a/cdc/kv/etcd_test.go
+++ b/cdc/kv/etcd_test.go
@@ -19,6 +19,9 @@ import (
 	"net/url"
 	"time"
 
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+
 	"github.com/pingcap/check"
 	"github.com/pingcap/errors"
 	"github.com/pingcap/ticdc/cdc/model"
@@ -261,14 +264,25 @@ func (s *etcdSuite) TestSetChangeFeedStatusTTL(c *check.C) {
 	c.Assert(status, check.DeepEquals, &model.ChangeFeedStatus{
 		ResolvedTs: 1,
 	})
-	err = s.client.SetChangeFeedStatusTTL(ctx, "test1", 1 /* seconds */)
+	err = s.client.SetChangeFeedStatusTTL(ctx, "test1", 1 /* second */)
 	c.Assert(err, check.IsNil)
 	status, _, err = s.client.GetChangeFeedStatus(ctx, "test1")
 	c.Assert(err, check.IsNil)
 	c.Assert(status, check.DeepEquals, &model.ChangeFeedStatus{
 		ResolvedTs: 1,
 	})
-	time.Sleep(1.5 * time.Second)
-	_, _, err = s.client.GetChangeFeedStatus(ctx, "test1")
-	c.Assert(errors.Cause(err), check.Equals, model.ErrChangeFeedNotExists)
+	for i := 0; i < 50; i++ {
+		_, _, err = s.client.GetChangeFeedStatus(ctx, "test1")
+		log.Warn("nil", zap.Error(err))
+		switch errors.Cause(err) {
+		case model.ErrChangeFeedNotExists:
+			return
+		case nil:
+			time.Sleep(100 * time.Millisecond)
+			continue
+		default:
+			c.Fatal("got unexpected error", err)
+		}
+	}
+	c.Fatal("the change feed status is still exists after 5 seconds")
 }

--- a/cdc/kv/etcd_test.go
+++ b/cdc/kv/etcd_test.go
@@ -249,3 +249,26 @@ func (s *etcdSuite) TestPutAllChangeFeedStatus(c *check.C) {
 		}
 	}
 }
+
+func (s *etcdSuite) TestSetChangeFeedStatusTTL(c *check.C) {
+	ctx := context.Background()
+	err := s.client.PutChangeFeedStatus(ctx, "test1", &model.ChangeFeedStatus{
+		ResolvedTs: 1,
+	})
+	c.Assert(err, check.IsNil)
+	status, _, err := s.client.GetChangeFeedStatus(ctx, "test1")
+	c.Assert(err, check.IsNil)
+	c.Assert(status, check.DeepEquals, &model.ChangeFeedStatus{
+		ResolvedTs: 1,
+	})
+	err = s.client.SetChangeFeedStatusTTL(ctx, "test1", 2 /* seconds */)
+	c.Assert(err, check.IsNil)
+	status, _, err = s.client.GetChangeFeedStatus(ctx, "test1")
+	c.Assert(err, check.IsNil)
+	c.Assert(status, check.DeepEquals, &model.ChangeFeedStatus{
+		ResolvedTs: 1,
+	})
+	time.Sleep(3 * time.Second)
+	_, _, err = s.client.GetChangeFeedStatus(ctx, "test1")
+	c.Assert(errors.Cause(err), check.Equals, model.ErrChangeFeedNotExists)
+}

--- a/cdc/owner.go
+++ b/cdc/owner.go
@@ -443,8 +443,10 @@ func (o *Owner) handleAdminJob(ctx context.Context) error {
 		o.adminJobs = o.adminJobs[removeIdx:]
 		o.adminJobsLock.Unlock()
 	}()
+	var deletedIdx int
 	for i, job := range o.adminJobs {
 		log.Info("handle admin job", zap.String("changefeed", job.CfID), zap.Stringer("type", job.Type))
+		deletedIdx = i
 		switch job.Type {
 		case model.AdminStop:
 			// update ChangeFeedDetail to tell capture ChangeFeedDetail watcher to cleanup
@@ -506,8 +508,8 @@ func (o *Owner) handleAdminJob(ctx context.Context) error {
 				return errors.Trace(err)
 			}
 		}
-		removeIdx = i + 1
 	}
+	removeIdx = deletedIdx + 1
 	return nil
 }
 

--- a/cdc/owner.go
+++ b/cdc/owner.go
@@ -450,7 +450,8 @@ func (o *Owner) handleAdminJob(ctx context.Context) error {
 			// update ChangeFeedDetail to tell capture ChangeFeedDetail watcher to cleanup
 			cf, ok := o.changeFeeds[job.CfID]
 			if !ok {
-				return errors.Errorf("changefeed %s not found in owner cache", job.CfID)
+				log.Warn("invalid admin job, changefeed not found", zap.String("changefeed", job.CfID))
+				continue
 			}
 			cf.info.AdminJobType = model.AdminStop
 			err := o.etcdClient.SaveChangeFeedInfo(ctx, cf.info, job.CfID)
@@ -475,10 +476,18 @@ func (o *Owner) handleAdminJob(ctx context.Context) error {
 			}
 		case model.AdminResume:
 			cfStatus, _, err := o.etcdClient.GetChangeFeedStatus(ctx, job.CfID)
+			if errors.Cause(err) != model.ErrChangeFeedNotExists {
+				log.Warn("invalid admin job, changefeed not found", zap.String("changefeed", job.CfID))
+				continue
+			}
 			if err != nil {
 				return errors.Trace(err)
 			}
 			cfInfo, err := o.etcdClient.GetChangeFeedInfo(ctx, job.CfID)
+			if errors.Cause(err) != model.ErrChangeFeedNotExists {
+				log.Warn("invalid admin job, changefeed not found", zap.String("changefeed", job.CfID))
+				continue
+			}
 			if err != nil {
 				return errors.Trace(err)
 			}

--- a/cdc/owner.go
+++ b/cdc/owner.go
@@ -443,7 +443,7 @@ func (o *Owner) handleAdminJob(ctx context.Context) error {
 		o.adminJobs = o.adminJobs[removeIdx:]
 		o.adminJobsLock.Unlock()
 	}()
-	var deletedIdx int
+	deletedIdx := -1
 	for i, job := range o.adminJobs {
 		log.Info("handle admin job", zap.String("changefeed", job.CfID), zap.Stringer("type", job.Type))
 		deletedIdx = i

--- a/cdc/owner.go
+++ b/cdc/owner.go
@@ -443,10 +443,9 @@ func (o *Owner) handleAdminJob(ctx context.Context) error {
 		o.adminJobs = o.adminJobs[removeIdx:]
 		o.adminJobsLock.Unlock()
 	}()
-	deletedIdx := -1
 	for i, job := range o.adminJobs {
 		log.Info("handle admin job", zap.String("changefeed", job.CfID), zap.Stringer("type", job.Type))
-		deletedIdx = i
+		removeIdx = i + 1
 		switch job.Type {
 		case model.AdminStop:
 			// update ChangeFeedDetail to tell capture ChangeFeedDetail watcher to cleanup
@@ -509,7 +508,6 @@ func (o *Owner) handleAdminJob(ctx context.Context) error {
 			}
 		}
 	}
-	removeIdx = deletedIdx + 1
 	return nil
 }
 

--- a/cdc/owner.go
+++ b/cdc/owner.go
@@ -477,7 +477,7 @@ func (o *Owner) handleAdminJob(ctx context.Context) error {
 			}
 		case model.AdminResume:
 			cfStatus, _, err := o.etcdClient.GetChangeFeedStatus(ctx, job.CfID)
-			if errors.Cause(err) != model.ErrChangeFeedNotExists {
+			if errors.Cause(err) == model.ErrChangeFeedNotExists {
 				log.Warn("invalid admin job, changefeed not found", zap.String("changefeed", job.CfID))
 				continue
 			}

--- a/cdc/owner.go
+++ b/cdc/owner.go
@@ -485,7 +485,7 @@ func (o *Owner) handleAdminJob(ctx context.Context) error {
 				return errors.Trace(err)
 			}
 			cfInfo, err := o.etcdClient.GetChangeFeedInfo(ctx, job.CfID)
-			if errors.Cause(err) != model.ErrChangeFeedNotExists {
+			if errors.Cause(err) == model.ErrChangeFeedNotExists {
 				log.Warn("invalid admin job, changefeed not found", zap.String("changefeed", job.CfID))
 				continue
 			}

--- a/cdc/owner.go
+++ b/cdc/owner.go
@@ -449,6 +449,7 @@ func (o *Owner) dispatchJob(ctx context.Context, job model.AdminJob) error {
 	if err != nil {
 		return errors.Trace(err)
 	}
+	// TODO Closing the resource should not be here
 	err = cf.ddlHandler.Close()
 	log.Info("stop changefeed ddl handler", zap.String("changefeed id", job.CfID), util.ZapErrorFilter(err, context.Canceled))
 	err = cf.sink.Close()
@@ -493,6 +494,11 @@ func (o *Owner) handleAdminJob(ctx context.Context) error {
 
 			// remove changefeed info
 			err = o.etcdClient.DeleteChangeFeedInfo(ctx, job.CfID)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			// set ttl to changefeed status
+			err = o.etcdClient.SetChangeFeedStatusTTL(ctx, job.CfID, 24*3600 /*24 hours*/)
 			if err != nil {
 				return errors.Trace(err)
 			}

--- a/cmd/client_changefeed.go
+++ b/cmd/client_changefeed.go
@@ -18,6 +18,8 @@ import (
 	"strings"
 	"time"
 
+	"github.com/pingcap/log"
+
 	"github.com/pingcap/ticdc/pkg/config"
 
 	"github.com/google/uuid"
@@ -143,6 +145,9 @@ func newQueryChangefeedCommand() *cobra.Command {
 				taskStatus = append(taskStatus, captureTaskStatus{CaptureID: captureID, TaskStatus: status})
 			}
 			meta := &cfMeta{Info: info, Status: status, Count: count, TaskStatus: taskStatus}
+			if info == nil {
+				log.Warn("this changefeed has been deleted, the residual meta data will be completely deleted within 24 hours.")
+			}
 			return jsonPrint(cmd, meta)
 		},
 	}

